### PR TITLE
[FIRRTL] Add the FIRRTL RefType Ops

### PR DESCRIPTION
**Pull request by prithayan**
This PR introduces two operations to handle `RefType`.
The `RefType` can be used to represent dataflow across modules and only module/instance ports can declare values of `RefType`.
The following operations are added to `FIRRTL` to express the dataflow across modules.

- RefSendOp
  - Send a read-only handle of a local value to a reference port.
  - Has connect-like semantics, but can only connect a `BaseType` value to a `RefType` port.
  - Expresses the source of a dataflow by getting it's reference. 
- RefResolveOp
  - Reads the `BaseType` value from a reference port
  - Represents the sink of a dataflow from the reference ports
  - The `RefType` is constrained to express a "single source to one-or-more sink" dataflow 
  - Hence, multiple `RefResolveOp`s can be remotely connected to a single `RefSendOp`.
  - If the Reference ops are lowered to cross-module-reference, then it represents the location of the XMR. 

Added few lit tests to verify proper errors for the constraints on the Reference ops and added the `xmr.mlir` to demonstrate the usage of Reference ops, it does not check for any output, just that there are no errors. `xmr.mlir` can possibly be removed later when other passes handling the Reference ops are introduced.

Co-authored-by: Will Dietz <will.dietz@sifive.com>

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 1]
```diff
@@ -662,3 +662,70 @@ def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
...
+//===----------------------------------------------------------------------===//
+
+class RefTypeConstraint<string ref, string base>
+  : TypesMatchWith<"reference base type should match",
+                   ref, base,
+                   "$_self.cast<RefType>().getType()">;
+
+class RefSingleUserConstraint<string ref>
+  : PredOpTrait<"reference port operand cannot be reused by any other op",
+                CPred<"$" # ref # ".hasOneUse()">>;
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:687]: `ref` is always `"ref"` right now. Would it be better to make this not take the parameter? Ditto for other constraints below.

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 2]
```diff
@@ -662,3 +662,70 @@ def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
...
+  let assemblyFormat = "$ref attr-dict `:` qualified(type($ref))";
+}
+
+def RefSendOp: FIRRTLOp<"ref.send", [RefTypeConstraint<"ref","result">,
+                                    RefSingleUserConstraint<"ref">,
+                                    RefPortConstraint<"ref">]> {
+  let summary = "FIRRTL Send through Reference";
+  let description = [{
+    Endpoint of a remote reference, send value through a reference
+    to be read from the firrtl.ref.resolve op.
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:725]:
```suggestion
    Endpoint of a remote reference. Send a value through a reference
    to be read from the firrtl.ref.resolve op.
```

File: test/Dialect/FIRRTL/errors.mlir [hunk 3]
```diff
@@ -1027,3 +1027,48 @@ firrtl.circuit "Top" {
...
+}
+
+// -----
+// Check flow semantics for ref.send
+
+firrtl.circuit "Foo" {
+  // expected-note @+1 {{destination was defined here}}
+  firrtl.module @Foo(in  %_a: !firrtl.ref<uint<1>>) {
+    %a = firrtl.wire : !firrtl.uint<1>
+    // expected-error @+1 {{connect has invalid flow: the destination expression "_a" has source flow, expected sink or duplex flow}}
```
Comment (seldridge) [inline test/Dialect/FIRRTL/errors.mlir:1048]: Nits: technically this is not a "connect", but a `ref.send`. Also, this can't have duplex flow (right now).

Fixing this is outside the scope of this PR.

File: test/Dialect/FIRRTL/errors.mlir [hunk 4]
```diff
@@ -1027,3 +1027,48 @@ firrtl.circuit "Top" {
...
+// Reference port cannot be reused
+
+firrtl.circuit "ForwardToInstance2x" {
+  firrtl.module @Bar2(out %_a: !firrtl.ref<uint<1>>) {
+    %zero = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.ref.send %_a, %zero : !firrtl.ref<uint<1>>
+  }
+  firrtl.module @Bar(out %_a: !firrtl.ref<uint<1>>) {
+    %x = firrtl.instance x @Bar2(out _a: !firrtl.ref<uint<1>>)
+    %y = firrtl.instance y @Bar2(out _a: !firrtl.ref<uint<1>>)
```
Comment (seldridge) [inline test/Dialect/FIRRTL/errors.mlir:1063]: Nit: this test can be golfed down slightly by inlining the `firrtl.ref.send` into module `@Bar`. The outer module `@ForwardToInstance2x` is then unnecessary. (Does the ref.send have to be paired with a ref.resolve?)

File: test/Dialect/FIRRTL/errors.mlir [hunk 5]
```diff
@@ -1027,3 +1027,48 @@ firrtl.circuit "Top" {
...
+    // expected-error @+1 {{connect operands of ref type cannot be reused}}
+    firrtl.strictconnect %_a, %x : !firrtl.ref<uint<1>>
+    firrtl.strictconnect %_a, %y : !firrtl.ref<uint<1>>
```
Comment (seldridge) [inline test/Dialect/FIRRTL/errors.mlir:1066]: We'll want to eventually relax this, but thanks for including a test to disallow it for now.

**Review by seldridge — APPROVED** [review]
Awesome to see this all coming together. 💯

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 6]
```diff
@@ -2042,8 +2042,11 @@ static LogicalResult checkRefTypeFlow(Operation *connect) {
               "connections only";
       return diag;
     }
-    if ((src.getType().isa<RefType>() && !src.hasOneUse()) ||
-        (dst.getType().isa<RefType>() && !dst.hasOneUse())) {
+    // RefType supports multiple readers. That is, there can be multiple
+    // RefResolveOps remotely connected to a single RefSendOp.
+    // But multiple RefSendOps should not be remotely connected to a single
+    // RefResolveOp.
+    if (dst.getType().isa<RefType>() && !dst.hasOneUse()) {
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:2049]: If this is just checking `dst`, probably can put this into the `if (dst-is-reftype)` above.

Do we still want this check?  It helps ensure `RefType`s aren't "sent" to multiple times, or connected to repeatedly, but also stops resolve from a send in same module, I think.  WDYT?

File: test/Dialect/FIRRTL/xmr.mlir [hunk 7]
```diff
@@ -0,0 +1,109 @@
+// RUN: circt-opt %s -split-input-file
```
Comment (dtzSiFive) [inline test/Dialect/FIRRTL/xmr.mlir:1]: Since we're moving away from "XMR" generally, should this be "ref.mlir" or something instead of "xmr.mlir"?
That said, looks like we're not losing the "XMR" association entirely so :shrug:.

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td (hunk 1 above)
[the remark this answers is not in the document]
Comment (seldridge) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:687]: I would like the ability to have a send/resolve in the same module. 👍

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 8]
```diff
@@ -662,3 +662,70 @@ def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
...
+// RefOps
+//===----------------------------------------------------------------------===//
+
+def RefResolveOp: FIRRTLOp<"ref.resolve", [RefTypeConstraint<"ref","result">,
+                                    RefInstancePortConstraint<"ref">]> {
+  let summary = "FIRRTL Resolve a Reference";
+  let description = [{
+    Resolve a remote reference for reading a remote value.
+    It takes a RefType input and returns the corresponding BaseType value.
+    If an XMR is emitted for this reference, it will be at the location
```
[the remark this answers is not in the document]
Comment (seldridge) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:710]: It could be lowered to a normal port. This would be useful for replacing the existing BoringUtils API. Whether or not that now makes sense is unclear given that the construction of ref ports is basically the same as the construction of ports.

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 9]
```diff
@@ -662,3 +662,70 @@ def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
 
   let assemblyFormat = "$input attr-dict `:` functional-type($input, $result)";
 }
+
+//===----------------------------------------------------------------------===//
+// RefOperations: Operations on the RefType.
+//
+// The current motivation is that these operations will be used to express
```
[the remark this answers is not in the document]
Comment (prithayan) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:669]: I agree, we should keep this correct for current situation. Let me rephrase it.

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td (hunk 1 above)
Comment (prithayan) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:687]: True, we can remove the `ref` from the parameters.

[the remark this answers is not in the document]
Comment (prithayan) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:687]: I think the way we developed the `RefType`, it should  represent dataflow across modules. If we only allow ports to be declared with `RefType`, then only `RefSend` and `RefResolve` can be used to connect `RefType` ports to `BaseType` values. 
If we want to add send and resolve in the same module, that might require using `wire` as an input, and that means `WireOp` needs to be declared with `RefType`.

[the remark this answers is not in the document]
Comment (prithayan) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:687]: Also, the single user constraint ensures that the dataflow is always explicit, single source to one-or-more sinks. This means `RefType` always represents a legal dataflow. In our case, every `RefResolve` can be statically traced to a single `RefSend`, else, it should be an error.

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td (hunk 8 above)
[the remark this answers is not in the document]
Comment (prithayan) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:710]: I think one issue with our `Ref` world is that we are developing it for an immediate use case of xmr, but it could potentially be used for other non-xmr stuffs which we have not considered in this design. This makes it a bit tricky to generalize the ops without the exact use cases. ;)

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 6 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:2049]: I agree, let me move this up into the if.
You are right, that this ensures every `RefResolve` can be traced to a single `RefSend` op. But allows the multiple read case. I am not sure, both send and resolve in a single module is possible. Other constraints (downward only reference) might not allow it, unless we declare a `wire` of `RefType`. Can you come up with an example for both send/resolve in one module ?

File: test/Dialect/FIRRTL/errors.mlir (hunk 3 above)
Comment (prithayan) [reply test/Dialect/FIRRTL/errors.mlir:1048]: We are reusing the `connect` verifier, hence the error message. Let's see if we can improve it.

File: test/Dialect/FIRRTL/errors.mlir (hunk 4 above)
Comment (prithayan) [reply test/Dialect/FIRRTL/errors.mlir:1063]: True, let me do that.

File: test/Dialect/FIRRTL/errors.mlir (hunk 5 above)
Comment (prithayan) [reply test/Dialect/FIRRTL/errors.mlir:1066]: this constraint ensures that every resolve can be statically traced to a single send. We can always express conditional XMR with a `when` on the resolve op.

File: test/Dialect/FIRRTL/xmr.mlir (hunk 7 above)
Comment (prithayan) [reply test/Dialect/FIRRTL/xmr.mlir:1]: Lets do it `ref.mlir`

File: test/Dialect/FIRRTL/errors.mlir (hunk 5 above)
Comment (seldridge) [reply test/Dialect/FIRRTL/errors.mlir:1066]: I'm paging this back in. Right, it seems like we don't want wire here, but to have `ref.send` return an SSA value. I don't think it makes sense to expand this to wires having `ref` type. Outside the scope of this PR.
